visidata/apps/vdsql/_ibis.py

```diff
@@ -80,24 +81,19 @@ def open_vdsql(vd, p, filetype=None):
 class IbisConnectionPool:
     def __init__(self, source, pool=None, total=0):
         self.source = source
-        self.pool = pool if pool is not None else []
-        self.total = total
+        self._local = threading.local()
```

Comment:
There's no need to store this on a member variable, you can just use `threading.local()` where it's needed.

Reply:
But this would be a good place to initialize `local().connection = None`.

Reply:
Hm, it doesn't. I'm not sure why `.connection` isn't sticky.:

```
<_thread._local object at 0x7f6c5852c090>
<_thread._local object at 0x7f6c5852c090>

```

Reply:
I think I know what's going on.

Reply:
Okay, so my guess was right. When we go into `getconn`, `threading.local()` will actually create a new thread-local object. So we'll need to assign it to a member variable.

Reply:
Sorry, I'm still trying to wrap my head around how thread-local storage works. It seems when we initialise self._local.connection = None in the constructor, that only sets the connection attribute for the main thread that created the IbisConnectionPool instance.
When a different thread tries to access `self._local.connection`, it's looking at its own thread-specific storage, which doesn't have the connection attribute initialized yet, and we get an `AttributeError`.

My instinct is to use `hasattr` here.

Reply:
`threading.local()` did create new instances. =( So we'll need to use member variables.